Allow custom tokenization in LimeTextExplainer

gsireesh opened the pull request:
In reference to #192, this pull request adds functionality and tests that allow passing in a callable as the `split_expression` to `LimeTextExplainer`. The changes in this PR can be summarized as follows:

- Allowing passing functions to the `LimeTextExplainer` for custom tokenization. The callable should accept a string and return a list of tokens.
- Tests for the `IndexedString` class, which is where the `split_expression` eventually is used. 
- changing the regex splitting functionality to use non-capturing groups. This allows us not to build a `non_vocabulary` since the separation characters, as defined in the regex, no longer appear in the split list, and therefore no longer need to be filtered out. This also allows for the same kind of iteration between the callable tokenization and the regex tokenization. 

This will close #192.

echo66 commented:
@gsireesh if you add the following to your first message

```
close #192
```

the original issue will be closed once this PR is accepted.

gsireesh commented:
@marcotcr sorry this took so long, got a little sidetracked. I've added a function to segment the string around the tokens when a tokenizer is passed in, and that seems to work to fix the `inverse_removing` breakage that you talked about last time. I've also added some tests for the `inverse_removing` function.

marcotcr commented:
Thanks for making the changes, and for the tests. I'm sorry I took so long to review this.

ryuzakace commented:
Hi! I am trying to use BERT tokenizer, but I am getting error - 

`Tokenization produced tokens that do not belong in string!`

`explainer =  LimeTextExplainer(
        split_expression=tokenizer.tokenize,
        bow=False,
        class_names=['Incorrect', 'Correct']
        )`

`exp = explainer.explain_instance(text[0],mod_ev, num_features=10)`


For string - `'I sent she away.'`

tokenizer.tokenize returns - `['i', 'sent', 'she', 'away', '.']`
